In [2]:
# Importar librerias

import pandas as pd
import matplotlib.pyplot as plt

# Análisis de estación meteorológica

## Importar archivo csv y conocer el tipo de dato

El archivo de datos se delimita con punto y coma, pero por defecto,  pd.read_csv() usa una coma (,) como delimitador.

In [3]:
estacion_meteorologica = pd.read_csv('datos_ ema_FPUNA.csv', sep=';')
print(type(estacion_meteorologica))

<class 'pandas.core.frame.DataFrame'>


## Nombre de las columnas, cantidad de filas y columnas, y tipo de dato de las columnas

In [4]:
estacion_meteorologica_nombres_columnas = estacion_meteorologica.columns
print("Nombre de las columnas: \n", estacion_meteorologica_nombres_columnas)

numero_filas, numero_columnas = estacion_meteorologica.shape
print("Numero de filas: ", numero_filas, "\nNumero de columnas: ", numero_columnas)

print("Tipo de dato de las columnas:\n", estacion_meteorologica.dtypes, sep="")

Nombre de las columnas: 
 Index(['Date / Time', 'veloc vto (m/s)', 'Direcc Vto (grados)',
       'Presion Atm (hPa)', 'Radic Global (w/m2)', 'humedad relativa',
       'Preciptacion (mm)', 'Temp del Aire °C', 'Veloc de Vto Max (m/s)'],
      dtype='object')
Numero de filas:  48084 
Numero de columnas:  9
Tipo de dato de las columnas:
Date / Time                object
veloc vto (m/s)           float64
Direcc Vto (grados)       float64
Presion Atm (hPa)         float64
Radic Global (w/m2)        object
humedad relativa          float64
Preciptacion (mm)         float64
Temp del Aire °C          float64
Veloc de Vto Max (m/s)    float64
dtype: object


La columna `Dat / Time` y `Radic Global (w/m2)` son de tipo objeto, por lo que se transformarán

In [5]:
# Convertir 'Date / Time' a tipo datetime, nuestro dato es DD/MM/YYYY HH:MM
estacion_meteorologica['Date / Time'] = pd.to_datetime(
    estacion_meteorologica['Date / Time'], 
    dayfirst=True, 
    errors='coerce'
)

# Convertir 'Radic Global (w/m2)' a tipo float, manejando valores no convertibles
estacion_meteorologica['Radic Global (w/m2)'] = pd.to_numeric(
    estacion_meteorologica['Radic Global (w/m2)'], 
    errors='coerce'
)

# Verificar los cambios
print(estacion_meteorologica.dtypes)


Date / Time               datetime64[ns]
veloc vto (m/s)                  float64
Direcc Vto (grados)              float64
Presion Atm (hPa)                float64
Radic Global (w/m2)              float64
humedad relativa                 float64
Preciptacion (mm)                float64
Temp del Aire °C                 float64
Veloc de Vto Max (m/s)           float64
dtype: object


## Estadística descriptiva del conjunto de datos

In [6]:
estacion_meteorologica.describe()

Date / Time  veloc vto (m/s)  Direcc Vto (grados)  \
count                          47752     47878.000000         47576.000000   
mean   2022-05-17 13:28:31.291673856         1.419218           150.294855   
min              2021-12-01 00:10:00         0.100000             0.000000   
25%              2022-02-22 18:27:30         0.700000            70.000000   
50%              2022-05-17 16:55:00         1.300000           157.000000   
75%              2022-08-09 09:12:30         2.000000           209.000000   
max              2022-10-31 21:40:00         7.700000           357.000000   
std                              NaN         0.970870            84.994602   

       Presion Atm (hPa)  Radic Global (w/m2)  humedad relativa  \
count       47340.000000         47569.000000      47239.000000   
mean          998.730042           199.510564         71.307818   
min             0.000000            -1.700000         17.000000   
25%           995.100000             0.700000         54.000000   
50%           998.600000             3.500000         73.000000   
75%          1002.100000           344.100000         91.000000   
max          1018.800000          1197.000000        100.000000   
std             6.940906           298.853920         21.141501   

       Preciptacion (mm)  Temp del Aire °C  Veloc de Vto Max (m/s)  
count       47842.000000      47620.000000            46987.000000  
mean            0.026759         22.924353                3.346700  
min             0.000000          2.100000                0.100000  
25%             0.000000         17.600000                1.600000  
50%             0.000000         22.600000                3.100000  
75%             0.000000         28.000000                4.700000  
max            25.600000         43.300000               19.600000  
std             0.388993          7.531531                2.235738

### Acceder a una columna (serie) de interés

In [7]:
temperatura_estacion_meteorologica = estacion_meteorologica['Temp del Aire °C']
print(temperatura_estacion_meteorologica)

# Conocer una fila específica
print('Fila 4: ', temperatura_estacion_meteorologica[4])


0        21.5
1        21.5
2        21.9
3        21.9
4        21.6
         ... 
48079    16.2
48080    15.4
48081    14.9
48082    14.9
48083    14.7
Name: Temp del Aire °C, Length: 48084, dtype: float64
Fila 4:  21.6


Conocer la media de una columna de interes

In [8]:
media_temperatura_estacion_meteorologica = temperatura_estacion_meteorologica.mean()
print("La media es: ", media_temperatura_estacion_meteorologica)

La media es:  22.924353212935745


Contar la cantidad de datos que tiene la columna de interes

In [9]:
cantidad_datos_temperatura_estacion_meteorologica = temperatura_estacion_meteorologica.count()
print(cantidad_datos_temperatura_estacion_meteorologica)

47620


### Conocer los primeros cinco valores del conjunto de datos

In [10]:
estacion_meteorologica.head()

Date / Time  veloc vto (m/s)  Direcc Vto (grados)  \
0 2021-12-01 00:10:00              0.6                197.0   
1 2021-12-01 00:20:00              0.4                192.0   
2 2021-12-01 00:30:00              0.5                182.0   
3 2021-12-01 00:40:00              0.7                175.0   
4 2021-12-01 00:50:00              0.6                183.0   

   Presion Atm (hPa)  Radic Global (w/m2)  humedad relativa  \
0              998.3                  1.9              68.0   
1              998.3                  1.9              68.0   
2              998.4                  1.9              64.0   
3              998.6                  1.0              65.0   
4              998.6                  0.8              67.0   

   Preciptacion (mm)  Temp del Aire °C  Veloc de Vto Max (m/s)  
0                0.0              21.5                     1.6  
1                0.0              21.5                     2.4  
2                0.0              21.9                     1.7  
3                0.0              21.9                     2.1  
4                0.0              21.6                     1.9

## Seleccionar datos específicos de nuestro conjunto de datos

In [11]:
#Primera fila
estacion_meteorologica.iloc[0]

Date / Time               2021-12-01 00:10:00
veloc vto (m/s)                           0.6
Direcc Vto (grados)                     197.0
Presion Atm (hPa)                       998.3
Radic Global (w/m2)                       1.9
humedad relativa                         68.0
Preciptacion (mm)                         0.0
Temp del Aire °C                         21.5
Veloc de Vto Max (m/s)                    1.6
Name: 0, dtype: object

In [12]:
# Primera fila y primera columna
estacion_meteorologica.iloc[0,0]

Timestamp('2021-12-01 00:10:00')

In [13]:
# Rango de fila: fila 1 a 2
estacion_meteorologica.iloc[1:3]

Date / Time  veloc vto (m/s)  Direcc Vto (grados)  \
1 2021-12-01 00:20:00              0.4                192.0   
2 2021-12-01 00:30:00              0.5                182.0   

   Presion Atm (hPa)  Radic Global (w/m2)  humedad relativa  \
1              998.3                  1.9              68.0   
2              998.4                  1.9              64.0   

   Preciptacion (mm)  Temp del Aire °C  Veloc de Vto Max (m/s)  
1                0.0              21.5                     2.4  
2                0.0              21.9                     1.7

In [14]:
# Rango de fila y columna: fila de 1 a 2 y columna de 0 a 1
estacion_meteorologica.iloc[1:3, 0:2]

Date / Time  veloc vto (m/s)
1 2021-12-01 00:20:00              0.4
2 2021-12-01 00:30:00              0.5

Seleccionar una fila específica por su índice

In [15]:
# Con la fila cero
fila_index_0 = estacion_meteorologica.loc[0]
fila_index_0

Date / Time               2021-12-01 00:10:00
veloc vto (m/s)                           0.6
Direcc Vto (grados)                     197.0
Presion Atm (hPa)                       998.3
Radic Global (w/m2)                       1.9
humedad relativa                         68.0
Preciptacion (mm)                         0.0
Temp del Aire °C                         21.5
Veloc de Vto Max (m/s)                    1.6
Name: 0, dtype: object

In [16]:
# Selecciona una columna por su nombre desde cero
presion_atm_estacion_meteorologica = estacion_meteorologica.loc[0:, 'Presion Atm (hPa)']
presion_atm_estacion_meteorologica

0         998.3
1         998.3
2         998.4
3         998.6
4         998.6
          ...  
48079    1008.2
48080    1008.2
48081    1008.3
48082    1008.2
48083    1008.2
Name: Presion Atm (hPa), Length: 48084, dtype: float64

In [17]:
# Selecciona dos columnas por su nombre desde cero
presionAtm_humedadRelativa_estacion_meteorologica = estacion_meteorologica.loc[0:, ['Presion Atm (hPa)', 'humedad relativa']]
presionAtm_humedadRelativa_estacion_meteorologica

Presion Atm (hPa)  humedad relativa
0                  998.3              68.0
1                  998.3              68.0
2                  998.4              64.0
3                  998.6              65.0
4                  998.6              67.0
...                  ...               ...
48079             1008.2              56.0
48080             1008.2              58.0
48081             1008.3              60.0
48082             1008.2              62.0
48083             1008.2              63.0

[48084 rows x 2 columns]

Conocer una tiempo (hora y fecha) en específico

In [18]:
hora_fecha = estacion_meteorologica.loc[estacion_meteorologica['Date / Time'] == "2021-12-01 00:10:00"]
hora_fecha

Date / Time  veloc vto (m/s)  Direcc Vto (grados)  \
0 2021-12-01 00:10:00              0.6                197.0   

   Presion Atm (hPa)  Radic Global (w/m2)  humedad relativa  \
0              998.3                  1.9              68.0   

   Preciptacion (mm)  Temp del Aire °C  Veloc de Vto Max (m/s)  
0                0.0              21.5                     1.6

Selecciona dos fechas (dos filas)

In [19]:
hora_fecha_2 = estacion_meteorologica.loc[
    (estacion_meteorologica['Date / Time'] == "2021-12-01 00:10:00") | 
    (estacion_meteorologica['Date / Time'] == "2021-12-01 00:20:00"),
    :
]
hora_fecha_2

Date / Time  veloc vto (m/s)  Direcc Vto (grados)  \
0 2021-12-01 00:10:00              0.6                197.0   
1 2021-12-01 00:20:00              0.4                192.0   

   Presion Atm (hPa)  Radic Global (w/m2)  humedad relativa  \
0              998.3                  1.9              68.0   
1              998.3                  1.9              68.0   

   Preciptacion (mm)  Temp del Aire °C  Veloc de Vto Max (m/s)  
0                0.0              21.5                     1.6  
1                0.0              21.5                     2.4

## Datos faltantes

In [20]:
datos_faltantes = estacion_meteorologica.isna()
datos_faltantes.head() # Mostrar los primeros cinco datos faltantes

Date / Time  veloc vto (m/s)  Direcc Vto (grados)  Presion Atm (hPa)  \
0        False            False                False              False   
1        False            False                False              False   
2        False            False                False              False   
3        False            False                False              False   
4        False            False                False              False   

   Radic Global (w/m2)  humedad relativa  Preciptacion (mm)  Temp del Aire °C  \
0                False             False              False             False   
1                False             False              False             False   
2                False             False              False             False   
3                False             False              False             False   
4                False             False              False             False   

   Veloc de Vto Max (m/s)  
0                   False  
1                   False  
2                   False  
3                   False  
4                   False

Sumar la cantidad de datos faltantes por columna

In [21]:
conteo_datos_faltantes = estacion_meteorologica.isna().sum()
conteo_datos_faltantes

Date / Time                332
veloc vto (m/s)            206
Direcc Vto (grados)        508
Presion Atm (hPa)          744
Radic Global (w/m2)        515
humedad relativa           845
Preciptacion (mm)          242
Temp del Aire °C           464
Veloc de Vto Max (m/s)    1097
dtype: int64

Existe una cantidad considerable de datos faltantes en ciertas variables

### Analizar la cantidad de datos faltantes

Calculo del porcentaje faltante para cada columna

In [22]:
porcentaje_datos_faltantes = estacion_meteorologica.isna().mean()*100
porcentaje_datos_faltantes

Date / Time               0.690458
veloc vto (m/s)           0.428417
Direcc Vto (grados)       1.056484
Presion Atm (hPa)         1.547292
Radic Global (w/m2)       1.071042
humedad relativa          1.757341
Preciptacion (mm)         0.503286
Temp del Aire °C          0.964978
Veloc de Vto Max (m/s)    2.281424
dtype: float64

Lo anterior indica que todas las columnas tienen menos de 3% de valores faltantes, lo cual es bajo.

### Rellenar información

La media para columnas con datos que no tienen valores atípicos significativos ni una fuerte dependencia temporal

In [23]:
# Rellenar con la media
columnas_media = ['Presion Atm (hPa)', 'Temp del Aire °C']
for col in columnas_media:
    estacion_meteorologica[col] = estacion_meteorologica[col].fillna(estacion_meteorologica[col].mean())

# Rellenar con la mediana
columnas_mediana = ['Preciptacion (mm)', 'Veloc de Vto Max (m/s)']
for col in columnas_mediana:    
    estacion_meteorologica[col] = estacion_meteorologica[col].fillna(estacion_meteorologica[col].median())

# Rellenar con la moda
columnas_moda = ['Direcc Vto (grados)']
for col in columnas_moda:
    moda = estacion_meteorologica[col].mode()[0]  # Obtener el primer valor de la moda
    estacion_meteorologica[col] = estacion_meteorologica[col].fillna(moda)

# Rellenar con interpolación lineal
columns_interpolate = ['Radic Global (w/m2)', 'humedad relativa', 'veloc vto (m/s)']
estacion_meteorologica[columns_interpolate] = estacion_meteorologica[columns_interpolate].interpolate(method='linear')


Verificar que no haya más valores faltantes

In [25]:
conteo_datos_faltantes = estacion_meteorologica.isna().sum()
conteo_datos_faltantes

Date / Time               332
veloc vto (m/s)             0
Direcc Vto (grados)         0
Presion Atm (hPa)           0
Radic Global (w/m2)         0
humedad relativa            0
Preciptacion (mm)           0
Temp del Aire °C            0
Veloc de Vto Max (m/s)      0
dtype: int64

Verificar si los valores de `Date / Time` están duplicados

In [31]:
duplicados_date_time_estacion_meteorologica = estacion_meteorologica['Date / Time'].duplicated().sum()
print(f'Número de duplicados en Date / Time: {duplicados_date_time_estacion_meteorologica}')

Número de duplicados en Date / Time: 331


In [32]:
duplicados_exactos = estacion_meteorologica[estacion_meteorologica.duplicated(subset='Date / Time', keep=False)]
print(duplicados_exactos)

      Date / Time  veloc vto (m/s)  Direcc Vto (grados)  Presion Atm (hPa)  \
143           NaT              0.5                175.0         998.730042   
287           NaT              1.3                144.0         998.730042   
431           NaT              1.3                 58.0         998.000000   
575           NaT              1.4                 58.0         998.730042   
719           NaT              1.3                183.0        1002.600000   
...           ...              ...                  ...                ...   
47377         NaT              1.7                259.0         992.500000   
47521         NaT              0.5                212.0         996.000000   
47665         NaT              0.1                140.0         992.300000   
47809         NaT              2.4                 59.0         988.700000   
47953         NaT              0.6                 87.0         988.500000   

       Radic Global (w/m2)  humedad relativa  Preciptacion (mm)

Hay 332 valores faltantes de la columna `Date / Time`, hay 331 filas duplicadas de esta columna.

Los 332 valores corresponde a NaT (not available) y se distribuyen de tal forma que 331 filas duplicadas tienen NaT y 1 fila única también tiene NaT